From b15cc4ff72165dc24e6d08f4e5ec3a936b079195 Mon Sep 17 00:00:00 2001
From: Maxim Mihaylin <slimbox.sdb@gmail.com>
Date: Sat, 5 Aug 2023 23:26:15 +0600
Subject: [PATCH] Changes dmenu behavior to show only desktop files. It is
 recommended to use the patch of the same name for dwm.

---
 Makefile           | 8 ++++++--
 dmenu_path_desktop | 7 +++++++
 dmenu_run_desktop  | 2 ++
 3 files changed, 15 insertions(+), 2 deletions(-)
 create mode 100755 dmenu_path_desktop
 create mode 100755 dmenu_run_desktop

diff --git a/Makefile b/Makefile
index a03a95c..752edc9 100644
--- a/Makefile
+++ b/Makefile
@@ -34,7 +34,7 @@ clean:
 dist: clean
 	mkdir -p dmenu-$(VERSION)
 	cp LICENSE Makefile README arg.h config.def.h config.mk dmenu.1\
-		drw.h util.h dmenu_path dmenu_run stest.1 $(SRC)\
+		drw.h util.h dmenu_path dmenu_path_desktop dmenu_run dmenu_run_desktop stest.1 $(SRC)\
 		dmenu-$(VERSION)
 	tar -cf dmenu-$(VERSION).tar dmenu-$(VERSION)
 	gzip dmenu-$(VERSION).tar
@@ -42,10 +42,12 @@ dist: clean
 
 install: all
 	mkdir -p $(DESTDIR)$(PREFIX)/bin
-	cp -f dmenu dmenu_path dmenu_run stest $(DESTDIR)$(PREFIX)/bin
+	cp -f dmenu dmenu_path dmenu_path_desktop dmenu_run dmenu_run_desktop stest $(DESTDIR)$(PREFIX)/bin
 	chmod 755 $(DESTDIR)$(PREFIX)/bin/dmenu
 	chmod 755 $(DESTDIR)$(PREFIX)/bin/dmenu_path
+	chmod 755 $(DESTDIR)$(PREFIX)/bin/dmenu_path_desktop
 	chmod 755 $(DESTDIR)$(PREFIX)/bin/dmenu_run
+	chmod 755 $(DESTDIR)$(PREFIX)/bin/dmenu_run_desktop
 	chmod 755 $(DESTDIR)$(PREFIX)/bin/stest
 	mkdir -p $(DESTDIR)$(MANPREFIX)/man1
 	sed "s/VERSION/$(VERSION)/g" < dmenu.1 > $(DESTDIR)$(MANPREFIX)/man1/dmenu.1
@@ -56,7 +58,9 @@ install: all
 uninstall:
 	rm -f $(DESTDIR)$(PREFIX)/bin/dmenu\
 		$(DESTDIR)$(PREFIX)/bin/dmenu_path\
+		$(DESTDIR)$(PREFIX)/bin/dmenu_path_desktop\
 		$(DESTDIR)$(PREFIX)/bin/dmenu_run\
+		$(DESTDIR)$(PREFIX)/bin/dmenu_run_desktop\
 		$(DESTDIR)$(PREFIX)/bin/stest\
 		$(DESTDIR)$(MANPREFIX)/man1/dmenu.1\
 		$(DESTDIR)$(MANPREFIX)/man1/stest.1
diff --git a/dmenu_path_desktop b/dmenu_path_desktop
new file mode 100755
index 0000000..e6b1144
--- /dev/null
+++ b/dmenu_path_desktop
@@ -0,0 +1,7 @@
+#!/bin/bash
+entries=$(ls /usr/share/applications/)
+for entry in $entries; do
+  if [[ $entry == *.desktop ]]; then
+    echo "${entry%.*}"
+  fi
+done;
diff --git a/dmenu_run_desktop b/dmenu_run_desktop
new file mode 100755
index 0000000..410b4b1
--- /dev/null
+++ b/dmenu_run_desktop
@@ -0,0 +1,2 @@
+#!/bin/sh
+dmenu_path_desktop | dmenu "$@" | ${SHELL:-"/bin/sh"} &
-- 
2.40.1

